pkg/obfuscate/shell_lexer.go

```diff
@@ -0,0 +1,327 @@
+	if token = scanner.Scan(expressions["doubleQuote"]); token != nil {
+		return &ShellToken{DoubleQuote, token.String(), pos, scanner.Index()}
+	}
+	if state.current == DoubleQuote {
+		var fullToken []string
+
+		for escaped := true; escaped; {
+			tokenStr := scanUntil(scanner, expressions["doubleQuote"])
+			if tokenStr == nil {
+				return nil
+			}
+
+			escaped = false
+			fullToken = append(fullToken, *tokenStr)
+			i := 1
+
+			for ; len(*tokenStr)-i > 0 && (*tokenStr)[len(*tokenStr)-i] == '\\'; i++ {
+				escaped = !escaped
+			}
+
+			if escaped {
+				fullToken = append(fullToken, "\"")
+				scanner.SetIndex(scanner.Index() + 1)
+			}
+		}
+
+		if len(fullToken) > 0 {
+			return &ShellToken{Field, strings.Join(fullToken, ""), pos, scanner.Index()}
+		}
+
+		return nil
+	}
+
+	// Handle single quoted strings
+	if token = scanner.Scan(expressions["singleQuote"]); token != nil {
+		return &ShellToken{SingleQuote, token.String(), pos, scanner.Index()}
+	}
+	if state.current == SingleQuote {
+		tokenString := scanUntil(scanner, expressions["singleQuote"])
+		if tokenString != nil {
+			return &ShellToken{Field, *tokenString, pos, scanner.Index()}
+		}
+		return nil
+	}
+
+	// General case
+	if token = scanner.Scan(expressions["parenthesesOpen"]); token != nil {
+		return &ShellToken{ParentheseOpen, token.String(), pos, scanner.Index()}
+	}
+
+	if token = scanner.Scan(expressions["parenthesesClose"]); token != nil {
+		return &ShellToken{ParentheseClose, token.String(), pos, scanner.Index()}
+	}
+
+	if token = scanner.Scan(expressions["dollar"]); token != nil {
+		return &ShellToken{Dollar, token.String(), pos, scanner.Index()}
+	}
+
+	if token = scanner.Scan(expressions["redirection"]); token != nil {
+		return &ShellToken{Redirection, token.String(), pos, scanner.Index()}
+	}
+
+	if token = scanner.Scan(expressions["equal"]); token != nil {
+		return &ShellToken{Equal, token.String(), pos, scanner.Index()}
+	}
+
+	if token = scanner.Scan(expressions["backticks"]); token != nil {
+		return &ShellToken{Backticks, token.String(), pos, scanner.Index()}
+	}
+
+	tokenString := scanUntil(scanner, expressions["anyToken"])
+	if tokenString != nil {
+		return &ShellToken{Field, *tokenString, pos, scanner.Index()}
+	}
+
+	return nil
+}
+
+func tokenize(token *ShellToken, state *State, ret []ShellToken) []ShellToken {
+	// Handle double quoted strings
+	if token.kind == DoubleQuote {
+		if state.current == DoubleQuote {
+			state.current = 0
+		} else {
+			state.current = DoubleQuote
+		}
+	}
+
+	// Handle single quoted strings
+	if token.kind == SingleQuote {
+		if state.current == SingleQuote {
+			state.current = 0
+		} else {
+			state.current = SingleQuote
+		}
+	}
+
+	// Handle shell subcommands
+	if token.kind == Field && len(ret) > 0 && ret[len(ret)-1].kind == DoubleQuote && strings.Contains(token.val, "$(") {
+		reg := regexp.MustCompile(`\$\(([^)]*)\)`)
+		matches := reg.FindAllStringSubmatchIndex(token.val, -1)
```

Comment:
You need to balance parenthesis (and manage quoted expressions). For instance, you could have to deal with multiple levels of subcommands. You can decide to accept some limitations and limite how deeply you want to parse, but you need to document it then :)

Reply:
> you need to document it then :)

I commented tests that are not supported, is that enough?